types/src/transaction/script.rs

```diff
@@ -50,3 +51,80 @@ impl fmt::Debug for Script {
             .finish()
     }
 }
+
+/// How to call a particular Move script (aka. an "ABI").
+#[derive(Clone, Debug, Hash, Eq, PartialEq, Serialize, Deserialize)]
+pub struct ScriptABI {
+    /// The public name of the script.
+    name: String,
+    /// Some text comment.
+    doc: String,
+    /// The `code` value to set in the `Script` object.
+    code: Vec<u8>,
+    /// The names of the type arguments.
+    ty_args: Vec<String>,
```

Comment:
It's probably worth having a `Vec<TypeArgumentABI>` here, where `TypeArgumentABI` is something like

```
pub struct TypeArgumentABI  {
   /// The name of the type argument
   name: String,
   /// The kind of the type argument: All, Copyable, or Resource. Could also use an enum instead of referencing the file_format:Kind
   kind_constraint: file_format::Kind
}
```

Reply:
Curious to hear what @vgao1996 / @dariorussi think here

Reply:
Sure, I could do that -- assuming that clients can indeed use the extra information. On this note, we will need to document type checking of type arguments (TypeTag vs Kind) and regular arguments (TransactionArgument vs TypeTag), so that clients know what to do.

p.s. TypeTag vs Kind: are there cases where we expect an infinite number of TypeTags? Otherwise, how about passing an enum of acceptable TypeTag values instead of kind?

Reply:
Yeah "ty_args: Vec<String>" seems a little bit weird to me and I'd expect "file_format::Kind" here instead. Also I'd use "ty_params" instead of "ty_args" here because this is a declaration. I'm unsure if we could have a `name: String` in `TypeArgumentABI ` because such information is not available in the file format. Do we have such annotations elsewhere?

Reply:
here I am assuming the `ty_args` refer to the generic type the script is providing so for `main<T>() {...}` it's the `T`. Am I correct?
Under that assumption it's true that the `Kind` is the only information needed by a client, and we can have an enum for that (whether using the `Kind` enum as is or defining a new one).
The name of the generic as in @sblackshear's `TypeArgumentABI` is an option if we wanted to do that.
The `String` really gives you nothing as the only info that is useful is the restrictions over the generic types.

Reply:
FYI I'm landing the PR with this definition to begin with:
```
pub struct TypeArgumentABI  {
   /// The name of the type argument
   name: String,
}
```